### PR TITLE
Terra 2.0 support

### LINKED ISSUE #2256: [Terra] Add support v2
1. Changing the “bank/MsgSend” to “cosmos-sdk/MsgSend”.
2. Wasm tx types need to be cosmwasm.wasm.v1.MsgExecuteContract.

https://github.com/terra-money/core/blob/main/WASM_MIGRATE.md#contract-address (edited)

### PR DESCRIPTION
**Pull request by catenocrypt**
## Description

Add support for Terra 2.0, alongside Terra Classic, in a backwards compatible way
(i.e.: using Terra Classic is the same as using Terra before, using Terra 2.0 requires some minimal changes compared to Terra). 
Fixes #2256 .

## How to test

Unit tests, both new and classic.
Real world transactions.

## Types of changes

* New feature (non-breaking change which adds functionality)

## Checklist

- [x] Create pull request as draft initially, unless its complete.
- [x] Add tests to cover changes as needed.
- [ ] Update documentation as needed.
- [x] If there is a related Issue, mention it in the description.
- [x] Add Terra 2 assets to assets repo
- [x] Update ios & kotlin tests
- [ ] Test staking messages

If you're adding a new blockchain

- [ ] I have read the [guidelines](https://developer.trustwallet.com/wallet-core/newblockchain#integration-criteria) for adding a new blockchain.

### CODE REVIEW

**Comment by catenocrypt** [comment]
Sample send transactions:
https://finder.terra.money/mainnet/tx/fbbe73ad2f0db3a13911dc424f8a34370dc4b7e8b66687f536797e68ee200ece
https://finder.terra.money/mainnet/tx/4F1A870F927FFD4CD42B6C5491A746B3C4334432A54735E62240CBE46394813A

**Comment by catenocrypt** [comment]
I was not yet been able to test token transfer/send transactions and contract execute transactions, there are no usable tokens on Terra 2.0 yet.

File: registry.json [hunk 1]
```diff
@@ -910,8 +910,37 @@
       "source": "https://github.com/terra-project/core",
       "rpc": "https://classic.fcd.terra.dev",
       "documentation": "https://docs.terra.money"
+    }
+  },
+  {
+    "id": "terra20",
```
Comment (hewigovens) [inline registry.json:916]:
```suggestion
    "id": "terrav2",
```

File: registry.json [hunk 2]
```diff
@@ -910,8 +910,37 @@
...
+    "name": "Terra20",
```
Comment (hewigovens) [inline registry.json:917]:
```suggestion
    "name": "TerraV2",
```

**Comment by hewigovens** [comment]
Terra v2 staking is available to test

File: src/Cosmos/ProtobufSerialization.h [hunk 3]
```diff
@@ -26,6 +26,10 @@ std::string buildProtoTxRaw(const Proto::SigningInput& input, const std::string&
 
 std::string buildProtoTxJson(const Proto::SigningInput& input, const std::string& serializedTx);
 
+nlohmann::json wasmExecuteTransferPayload(const Proto::Message_WasmExecuteContractTransfer& msg);
+
+nlohmann::json wasmExecuteSendPayload(const Proto::Message_WasmExecuteContractSend& msg);
+
 nlohmann::json wasmTerraExecuteTransferPayload(const Proto::Message_WasmTerraExecuteContractTransfer& msg);
```
Comment (hewigovens) [inline src/Cosmos/ProtobufSerialization.h:33]: Let's move terra / thorchain naming methods to another file?

File: swift/Podfile.lock [hunk 4]
```diff
@@ -13,4 +13,4 @@ SPEC CHECKSUMS:
 
 PODFILE CHECKSUM: aac2324ba35cdd5631cb37618cd483887bab9cfd
 
-COCOAPODS: 1.11.3
+COCOAPODS: 1.11.2
```
Comment (hewigovens) [inline swift/Podfile.lock:16]: please upgrade your installed cocoapods

File: android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/terra/TestTerraTransactions.kt [hunk 5]
```diff
@@ -99,111 +98,17 @@ class TestTerraTransactions {
...
         }.build()
 
-        val output = AnySigner.sign(signingInput, TERRA, SigningOutput.parser())
-
-        assertEquals(output.serialized, "{\"mode\":\"BROADCAST_MODE_BLOCK\",\"tx_bytes\":\"CucBCuQBCiYvdGVycmEud2FzbS52MWJldGExLk1zZ0V4ZWN1dGVDb250cmFjdBK5AQosdGVycmExOHd1a3A4NGRxMjI3d3U0bWdoMGptNm45bmxuajZyczgycHA5d2YSLHRlcnJhMTR6NTZsMGZwMmxzZjg2enkzaHR5Mno0N2V6a2hudGh0cjl5cTc2Glt7InRyYW5zZmVyIjp7ImFtb3VudCI6IjI1MDAwMCIsInJlY2lwaWVudCI6InRlcnJhMWpsZ2FxeTludm4yaGY1dDJzcmE5eWN6OHM3N3duZjlsMGttZ2NwIn19EmcKUApGCh8vY29zbW9zLmNyeXB0by5zZWNwMjU2azEuUHViS2V5EiMKIQNwZjrHsPmJKW/rXOWfukpQ1+lOHOJW3/IlFFnKLNmsABIECgIIARgDEhMKDQoFdWx1bmESBDMwMDAQwJoMGkAaprIEMLPH2HmFdwFGoaipb2GIyhXt6ombz+WMnG2mORBI6gFt0M+IymYgzZz6w1SW52R922yafDnn7yXfutRw\"}")
-        assertEquals(output.error, "")
-    }
-
-    @Test
-    fun testSigningWasmTerraGenericProtobuf() {
```
Comment (hewigovens) [inline android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/terra/TestTerraTransactions.kt:117]: move to luna classic test?

File: swift/Tests/Blockchains/TerraTests.swift [hunk 6]
```diff
@@ -1,4 +1,4 @@
-// Copyright © 2017-2020 Trust Wallet.
+// Copyright © 2017-2022 Trust Wallet.
```
Comment (hewigovens) [inline swift/Tests/Blockchains/TerraTests.swift:1]: move old tests for classic to a separate test file

**Comment by catenocrypt** [comment]
Note: Terra 2.0 asset info has been included in assets repo  https://github.com/trustwallet/assets/tree/master/blockchains/terrav2/info

File: android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/terra/TestTerraTransactions.kt (hunk 5 above)
Comment (catenocrypt) [reply android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/terra/TestTerraTransactions.kt:117]: I have removed the Terra classic specific tests (swift, kotlin), not relevant any more

File: tests/Terra/TWCoinTypeTests.cpp [hunk 7]
```diff
@@ -13,7 +13,27 @@
...
+    auto accUrl = WRAPS(TWCoinTypeConfigurationGetAccountURL(TWCoinTypeTerraV2, accId.get()));
+    auto id = WRAPS(TWCoinTypeConfigurationGetID(TWCoinTypeTerraV2));
+    auto name = WRAPS(TWCoinTypeConfigurationGetName(TWCoinTypeTerraV2));
+
+    ASSERT_EQ(TWCoinTypeConfigurationGetDecimals(TWCoinTypeTerraV2), 6);
+    ASSERT_EQ(TWBlockchainCosmos, TWCoinTypeBlockchain(TWCoinTypeTerraV2));
+    ASSERT_EQ(0x0, TWCoinTypeP2shPrefix(TWCoinTypeTerraV2));
+    ASSERT_EQ(0x0, TWCoinTypeStaticPrefix(TWCoinTypeTerraV2));
+    assertStringsEqual(symbol, "LUNA");
+    assertStringsEqual(txUrl, "https://finder.terra.money/tx/tx/D28D8AFC7CE89F2A22FA2DBF78D2C0A36E549BB830C4D9FA7459E3F723CA7182");
```
Comment (Milerius) [inline tests/Terra/TWCoinTypeTests.cpp:30]: should be `https://finder.terra.money/mainnet/tx/` instead of `https://finder.terra.money/tx/tx/`

Comment (hewigovens) [reply tests/Terra/TWCoinTypeTests.cpp:30]: good catch, I will fix it

**Review by hewigovens — APPROVED** [review]

**Review by Milerius — APPROVED** [review]

**Pull request merged**